Workflow/Sources/WorkflowObserver.swift

```diff
@@ -0,0 +1,262 @@
+    )
+
+    /// Indicates a `Workflow` produced its initial state value.
+    /// - Parameters:
+    ///   - workflow: The `Workflow` that just produced its initial state.
+    ///   - initialState: The `State` that was created.
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`
+    func workflowDidMakeInitialState<WorkflowType: Workflow>(
+        _ workflow: WorkflowType,
+        initialState: WorkflowType.State,
+        session: WorkflowSession
+    )
+
+    /// Called before a `Workflow` is queried for its latest `Rendering`.
+    /// - Parameters:
+    ///   - workflow: The `Workflow` that is about to be render.
+    ///   - state: The corresponding `State` that will be used during the render call.
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`.
+    /// - Returns: An optional closure to be called immediately after the new `Rendering` is produced, which takes the rendering as the only parameter.
+    func workflowWillRender<WorkflowType: Workflow>(
+        _ workflow: WorkflowType,
+        state: WorkflowType.State,
+        session: WorkflowSession
+    ) -> ((WorkflowType.Rendering) -> Void)?
+
+    /// Called after an existing `Workflow` is updated.
+    /// - Parameters:
+    ///   - oldWorkflow: The previous `Workflow`
+    ///   - newWorkflow: The new `Workflow`
+    ///   - state: The state **after** the update has occurred.
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`.
+    func workflowDidChange<WorkflowType: Workflow>(
+        from oldWorkflow: WorkflowType,
+        to newWorkflow: WorkflowType,
+        state: WorkflowType.State,
+        session: WorkflowSession
+    )
+
+    /// Called after a `WorkflowAction` is received, but **before** it has been handled and propagated up the tree.
+    /// - Parameters:
+    ///   - action: The action that was received.
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`.
+    func workflowDidReceiveAction<Action: WorkflowAction>(
+        _ action: Action,
+        session: WorkflowSession
+    )
+}
+
+// MARK: - WorkflowSession
+
+/// A `WorkflowSession`encapsulates the information that gives a `WorkflowNode` its identity.
+/// In particular, it captures:
+/// - The type of the corresponding `Workflow` conformance.
+/// - The `String` key used when the workflow was rendered.
+/// - An `Identifier` type that is unique across program execution.
+/// - An optional reference to a parent `WorkflowSession`, to differentiate root nodes.
+public struct WorkflowSession {
+    public struct Identifier: Hashable {
+        private static var _nextRawID: UInt64 = 0
+        private static func _makeNextSessionID() -> UInt64 {
+            _nextRawID += 1
+            return _nextRawID
+        }
+
+        let rawIdentifier: UInt64 = Self._makeNextSessionID()
+    }
+
+    private indirect enum IndirectParent {
```

Comment:
out of curiosity, what drove the need for this?

Reply:
structs and enums cannot contain stored properties that reference their type. originally i had this as a class, but swapped to the 'indirect enum case' (i.e. a custom optional implementation that allows recursion), mostly just to see if implementing this with the value types was possible.

IMO it is a bit unusual, but seems to work. though it might also make sense to remove the parent reference entirely if it doesn't seem valuable. it does let you walk from a given node up to the root though, which might be useful in some circumstances.

Reply:
To make sure I understand, if you deleted the `indirect` keyword from the declaration of `IndirectParent`, it wouldn't compile?

I see that is true if some `case` of `IndirectParent` has an associated value of type `IndirectParent`, but I didn't realize that's true also if the associated value type is only _mutually_ recursive with `IndirectParent`, like `WorkflowSession` is.

Reply:
> To make sure I understand, if you deleted the indirect keyword from the declaration of IndirectParent, it wouldn't compile?

@square-tomb that's right. if i remove it, i get the following compiler error at the declaration of the `_indirectParent` property:

> Value type 'WorkflowSession' cannot have a stored property that recursively contains it

i assume this is because value types are to be stack-allocated, and so their entire layout must have a fixed size (or at least a bounded size?) so the compiler knows how to set up the function prologues appropriately. if a value type contains a recursive reference (anywhere in a descendant property), then you get this error, presumably because the type's layout size cannot be determined.

> Another typical workaround is to use a "box"

@JaviSoto yep, and i believe that is essentially what's done internally when you mark an enum as `indirect`. i found [this thread](https://forums.swift.org/t/who-benefits-from-the-indirect-keyword/20167) had some useful information and links on the topic.